### ISSUE TITLE
[no-unused-var] varsIgnorePattern not working for this module definition

### ISSUE THREAD

**Issue by HolgerJeromin** [typescript-eslint/typescript-eslint#2648]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JS
module.exports = {
    "root": true,
    "parser": "@typescript-eslint/parser",
    "plugins": [        "@typescript-eslint"    ],
    "rules": {
        "no-unused-vars": "off",
        "@typescript-eslint/no-unused-vars": ["error", {
            "vars": "all",
            "varsIgnorePattern": "Hmi",
            "args": "none"
        }]
    }

}
```

```TS
module Hmi {
    export interface myType  {
    }
    export function getType(): Hmi.myType | null {
        return null; 
     // return Hmi.System.getMyType();
    }
}
```
Or even shorter:
````ts
module Hmi {
    export class myClass {
    }
    console.log(Hmi.myClass);  // Register call in reality
}
````

**Expected Result**

No rule violation. 
1. ~~Because the module is used.~~
2. Because I even added the name `Hmi` into `varsIgnorePattern`

**Actual Result**

````
.\node_modules\.bin\eslint.cmd  file.ts

file.ts
  1:8  error  'Hmi' is defined but never used  @typescript-eslint/no-unused-vars

✖ 1 problem (1 error, 0 warnings)
````

**Additional Info**
This examples are working!

```TS
export module Hmi {   // <-- export added!
    export interface myType  {
    }
    export function getType(): Hmi.myType | null {
        return null;
    }
}
```
```TS
module Hmi {
    export interface myType  {
    }
    export function getType():  null {  // <-- myType removed
        return null;
    }
}
```

**Versions**

      "@typescript-eslint/eslint-plugin": "4.4.0",
      "@typescript-eslint/parser": "4.4.0",
      "eslint": "6.8.0", and "7.10.0"
      "typescript": "4.0.3",

**Comment by bradzacher** [comment]
> No rule violation.
> Because the module is used.

Your example code doesn't include a usage.
Please update your example appropriately

**Comment by HolgerJeromin** [comment]
This is a minimal example for the behavior in our 460000 loc project with TS compiler option `"module": "none"`.
I adjusted the examples a bit.
I added a `console` call to the example which is a register call in reality. 
But in my code I really have files which are only such wrappers of an internal API. They are called from customer code.

It is fine if eslint can not detect usage of the object. But at least the entry `varsIgnorePattern` should make him happy.

**Comment by bradzacher** [comment]
I ask because the usage is ***very*** important.
When you provide an example as part of a bug report, that example needs to be a self-contained reproduction of the issue. If it's not, then there's no way I can repro, and thus no way I can help you.
So thank you for updating.

In this case - in your updated example...

Usages from within the namespace do not count, and are not considered by the rule.
This is done on purpose. Or else you could have orphaned, self-referencing namespaces left in your codebase.

You get the same functionality with `function`s:
```ts
/* eslint no-unused-vars: ["error"] */

function foo() {
  return foo(); // Error: 'foo' is defined but never used. (no-unused-vars)
}
```
[eslint demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVudXNlZC12YXJzOiBbXCJlcnJvclwiXSAqL1xuXG5mdW5jdGlvbiBmb28oKSB7XG4gIHJldHVybiBmb28oKTsgXG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMiwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6eyJqc3giOnRydWV9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

**Comment by WeebHiroyuki** [comment]
I can reproduce this issue with an example like this:
```ts
interface Person {
  name: string;
  age: number;
  userID: number;
}

const John: Person = {
  name: 'John',
  age: 25,
  userID: 5031
};
```
typescript-eslint would end up saying that `Person` was never used.

**Comment by bradzacher** [comment]
@WeebHiroyuki please file a new issue. Your issue is unrelated.

**Comment by HolgerJeromin** [comment]
> Usages from within the namespace do not count, and are not considered by the rule.
> This is done on purpose. Or else you could have orphaned, self-referencing namespaces left in your codebase.

Ok, I am perfectly fine with that. 

Another example from my code base:

````ts
module Hmi.System { 
    export class Init { 
        public static run() {
             // Run the HMI
        }
    }
    Hmi.System.Init.run(); // This code kick starts the main code.
}
````

I have no problem adding `"varsIgnorePattern": "Hmi"` at the main eslint config. 
But starting every ts file with 
`module Hmi { // eslint-disable-line @typescript-eslint/no-unused-vars` 
is not very sexy.

**Comment by bradzacher** [comment]
I guess I have a few questions here before I can provide a good course of action:
- why are you using module/namespace syntax at all?
- why are you using classes as static function containers? (instead of an object, or another namespace)
- why are you nesting the run call within the namespace scope, instead of the parent (root) scope?

**Comment by HolgerJeromin** [comment]
> why are you using module/namespace syntax at all?
> why are you using classes as static function containers? (instead of an object, or another namespace)

It is a technical dept in our code (which started with TS 1.6) we want to change in the (far?) future.

> why are you nesting the run call within the namespace scope, instead of the parent (root) scope?

You are right. This can be adjusted with some files (but not in all).

Sorry if I repeat myself. But I see `varsIgnorePattern` as a flag like "whatever the code is structured this var should never ever be reported".

**Comment by HolgerJeromin** [comment]
> In this case, yes likely using varsIgnorePattern to ignore Hmi is the best course of action for your codebase.

I think you did not understood what I am trying to say since four messages:
`"varsIgnorePattern": "Hmi"` 
**Does not ignore the var `Hmi`!** in my example (and real) code.

**Comment by bradzacher** [comment]
I did not understand that was what you meant, no sorry.

The bug is here. Our extensions to rule doesn't consider `varsIgnorePattern` for namespace decls that are self-referenced

https://github.com/typescript-eslint/typescript-eslint/blob/c41dbe56e0514846e4d21fc5fcd8847da50e92c6/packages/eslint-plugin/src/rules/no-unused-vars.ts#L199-L209

Should be a simple fix. happy to accept a PR.
